videosink/display.go

```diff
@@ -0,0 +1,102 @@
+// Copyright 2021 The Periph Authors. All rights reserved.
+// Use of this source code is governed under the Apache License, Version 2.0
+// that can be found in the LICENSE file.
+
+// The videosink package provides a display driver implementing an HTTP request
+// handler. Client requests get an initial snapshot of the graphics buffer and
+// are updated further on every change.
+//
+// The primary use case is the development of display outputs on a host
+// machine. Additionally devices with network connectivity can use this driver
+// to provide a copy of their local display via a web interface.
+//
+// The protocol used is "MJPEG" (https://en.wikipedia.org/wiki/Motion_JPEG)
+// which is often used by IP cameras. Because of its better suitability for
+// computer-drawn graphics the PNG image format is used by default. JPEG as
+// a format can be selected via Options.Format or using the "format" URL
+// parameter.
+package videosink
+
+import (
+	"image"
+	"image/color"
+	"image/draw"
+	"net/http"
+	"sync"
+
+	"periph.io/x/conn/v3/display"
+)
+
+// Options for videosink devices.
+type Options struct {
+	// Width and height of the image buffer.
+	Width, Height int
+
+	// Format specifies the image format to send to clients.
+	Format ImageFormat
+}
+
+type Display struct {
+	defaultFormat ImageFormat
+
+	mu       sync.Mutex
+	buffer   *image.RGBA
+	clients  map[*client]struct{}
+	snapshot map[imageConfig][]byte
+}
+
+var _ display.Drawer = (*Display)(nil)
+var _ http.Handler = (*Display)(nil)
+
+// New creates a new videosink device instance.
+func New(opt Options) *Display {
+	buffer := image.NewRGBA(image.Rect(0, 0, opt.Width, opt.Height))
+	draw.Draw(buffer, buffer.Bounds(), image.Black, image.Point{}, draw.Src)
```

Comment:
It's unnecessary, it's the default.

Reply:
Not quite; the difference is in the alpha channel. `buffer.At(0, 0)` before that call to `draw.Draw` is `color.RGBA{R:0x0, G:0x0, B:0x0, A:0x0}`, after it's `color.RGBA{R:0x0, G:0x0, B:0x0, A:0xff}`. This doesn't matter for JPEG, but it does for PNG. Maybe we should leave this to the caller, however. What do you think?

Reply:
oh good point!